sycl/doc/extensions/proposed/sycl_ext_oneapi_enqueue_functions.asciidoc

```diff
@@ -0,0 +1,684 @@
+existing queue shortcuts. A future version of this extension may adjust this
+overload set to include functions compatible with future C++ concepts (e.g,
+by accepting a scheduler and returning a sender).
+
+
+=== Usage example
+
+The example below demonstrates that the syntax proposed here requires only
+minor changes to existing applications, while retaining their structure.
+
+
+==== SYCL 2020
+
+[source,c++]
+----
+q.submit([&](sycl::handler& h) {
+  sycl::accessor result { buf, h, sycl::write_only, sycl::no_init };
+  h.parallel_for(1024, [=](sycl::id<1> idx) {
+    result[idx] = idx;
+  });
+});
+
+float* output = sycl::malloc_shared<int>(1, q);
+*output = 0;
+std::vector<sycl::event> depEvents = /* some dependencies */;
+sycl::event e = q.parallel_for(sycl::nd_range<1>{1024, 16}, depEvents,
+  sycl::reduction(output, sycl::plus<>()),
+  [=](sycl::nd_item<1> it, auto& sum) {
+    sum += it.get_global_id();
+});
+e.wait();
+sycl::free(output, q);
+----
+
+
+==== Proposed syntax
+
+[source,c++]
+----
+using syclex = sycl::ext::oneapi::experimental;
+
+syclex::submit(q, [&](sycl::handler& h) {
+  sycl::accessor result { buf, h, sycl::write_only, sycl::no_init };
+  syclex::parallel_for(h, 1024, [=](sycl::id<1> idx) {
+    result[idx] = idx;
+  });
+});
+
+float* output = sycl::malloc_shared<int>(1, q);
+*output = 0;
+std::vector<sycl::event> depEvents = /* some dependencies */;
+sycl::event e = syclex::submit_with_event(q, [&](sycl::handler& h) {
+  h.depends_on(depEvents);
+  syclex::nd_launch(h, sycl::nd_range<1>{1024, 16},
+    [=](sycl::nd_item<1> it, auto& sum) {
+      sum += it.get_global_id();
+    },
+    sycl::reduction(output, sycl::plus<>())
+});
+e.wait();
+sycl::free(output, q);
+----
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_ENQUEUE_FUNCTIONS` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Launch configuration
+
+A launch configuration object of type `launch_config` is used to bundle
+together all components of a kernel's launch configuration, including:
+
+1. The range of execution.
+2. Any compile-time properties.
+
+[NOTE]
+====
+Any compile-time properties passed to a `launch_config` do not affect the
+type of the enqueued kernel, and are primarily intended to be used for
+properties with run-time values that may impact how a kernel is launched
+by an underlying backend.
```

Comment:
Is it really true that we primarily expect these properties to be runtime?  I think it would be common to pass `sub_group_size` here, and that's a compile-time property.  Note that `launch_config` is templated on `Properties`, which would not be necessary if the properties were always runtime values.

Reply:
> Is it really true that we primarily expect these properties to be runtime? I think it would be common to pass `sub_group_size` here, and that's a compile-time property.

The intent here was that properties which only affect the launch would go into the `launch_config`, whereas properties that (may) affect the code generated for the kernel would be attached to the kernel itself.  `sub_group_size` is an example of the latter case:

```c++
// sub-group size modifies the kernel function itself 
auto kernel = user_defined_wrapper([=]() { ... }, sycl::properties{ sycl::sub_group_size<8> });

// foo property with run-time value bar affects how the kernel function is launched
sycl::parallel_for(q, launch_config(1024, sycl::properties { sycl::foo(bar) }, kernel);
```

There's no reason to disallow properties with compile-time values from being passed to `launch_config`, but it would only be useful for optimizing the host application (e.g., to select a different overload of `parallel_for` or use something like an `if constexpr` inside the implementation).

> Note that `launch_config` is templated on `Properties`, which would not be necessary if the properties were always runtime values.

I don't think this is true.  The type of a `sycl::properties` reflects all of the properties in it, just not necessarily their values.  If a property with a run-time value is used to construct a `sycl::properties`, the type of the property is still one if its template arguments.

Reply:
> The intent here was that properties which only affect the launch would go into the `launch_config`, whereas properties that (may) affect the code generated for the kernel would be attached to the kernel itself. `sub_group_size` is an example of the latter case

This is different from the sycl_ext_oneapi_kernel_properties extension.  If we are intentionally changing this, we should state it more explicitly and in a normative way.  Maybe something like:

> The properties to `launch_config` only affect the way in which the kernel is launched.  They cannot be used to define information about the kernel itself.  This extension does not define any properties for `launch_config`, but other extensions could define such properties.
>
> [Note: The properties defined in [sycl_ext_oneapi_kernel_properties](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc) (e.g. `work_group_size`) cannot be used in `launch_config`.  In order to use these properties with a kernel, the kernel must be a named functioned object which exposes the properties via `get(sycl::ext::oneapi::experimental::properties_tag)` as described in that extension. -- end note]

Reply:
Good point, I'll make this change.

Eventually, I think we should update `sycl_ext_oneapi_kernel_properties` (so that it only defines the properties, not the `parallel_for` overloads).  Do you agree it makes sense to hold off on making that change until we've got feedback on this new extension?